CMSgov/bluebutton-web-server
# [BB2-1774] Add application fields for data access types

**JIRA Ticket:**
[BB2-1774](https://jira.cms.gov/browse/BB2-1774)

**User Story or Bug Summary:**


In order to implement data access expiration dates, new fields need to be added to the Application model.

This support the "Access Grant Token Limits" Jira epic work.


### What Does This PR Do?



The following is a summary of the changed files:

- apps/dot_ext/utils.py:
  - Created `is_data_access_type_valid()` function that:
    - Validates data_access_type & end_date combos.
    - Returns the error message to be used in a target exception by the caller.
  - Moved `Application = get_application_model()` in to scope of function where it is used, since this was causing an issue.

- apps/dot_ext/models.py:
  - New fields are added to the Application model:
    - `Application.data_access_type`
    - `Application.end_date`
  - A `save()` override method is added to the `Application` model.
    - Validation is performed:
      - Check if `data_access_type` is a valid choice.
      - Check for valid `data_access_type` & `end_date` combos using the `is_data_access_type_valid()` function from utils.py.

- apps/dot_ext/migrations/0004_auto_20221117_2012.py:
  - Created a migration script to add fields to the database.

- apps/dot_ext/admin.py:
  - New fields are added to the `MyApplications` Django admin.
  - Added to the `clean()` method to validate field entry when using the admin utilizing the `is_data_access_type_valid()` function from utils.py.
  - Added the new fields to the admin LIST DISPLAY.
  - Removed fields from the list_display, list_filter, search_files that were not useful/needed.
    - For example, client_type & authorization_grant_types only have one choice currently.

- apps/test.py:
  - Updated the `_create_application()` test helper method for the new fields.

- apps/dot_ext/tests/test_models.py:
  - Created a method for model tests: `TestDotExtModels.test_application_data_access_fields()` 
    - Added testing related to CRUD operations and field validations.



### What Should Reviewers Watch For?



If you're reviewing this PR, please check these things, in particular:

- Review of code style and namings.
- Testing of the new fields using the Django `MyApplications` admin.
- Review and test run of the Django tests.
  - Command to just run the single test:  
    ```
    docker-compose exec web python runtests.py apps.dot_ext.tests.test_models.TestDotExtModels.test_application_data_access_fields
    ```
  - Can you think of additional tests to add?

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies?  **No**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?  **No**.


### What Needs to Be Merged and Deployed Before this PR?



### Any Migrations?



  * [ ] There is a more complicated migration plan (downtime, etc) 


### Submitter Checklist



I have gone through and verified that...:

    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
    * Please review the standards every few months to ensure you're familiar with them.
